src/Middleware/ForceSecureConnection.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Web\Middleware;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Yiisoft\Http\Header;
+use Yiisoft\Http\Status;
+
+/**
+ * Redirects insecure requests from HTTP to HTTPS, and adds headers necessary to enhance security policy.
+ *
+ * HTTP Strict-Transport-Security (HSTS) header is added to each response and tells the browser that your site works on
+ * HTTPS only.
+ *
+ * The Content-Security-Policy (CSP) header can force the browser to load page resources only through a secure
+ * connection, even if links in the page layout are specified with an unprotected protocol.
+ *
+ * Note: Prefer forcing HTTPS via web server in case you are not creating installable product such as CMS and not
+ * hosting the project on a server where you do not have access to web server configuration.
+ */
+final class ForceSecureConnection implements MiddlewareInterface
+{
+    private const DEFAULT_CSP_DIRECTIVES = 'upgrade-insecure-requests; default-src https:';
+    private const DEFAULT_HSTS_MAX_AGE = 31_536_000; // 12 months
+
+    private bool $redirect = true;
+    private int $statusCode = Status::MOVED_PERMANENTLY;
+    private ?int $port = null;
+
+    private bool $addCSP = true;
+    private string $cspDirectives = self::DEFAULT_CSP_DIRECTIVES;
+
+    private bool $addHSTS = true;
+    private int $hstsMaxAge = self::DEFAULT_HSTS_MAX_AGE;
+    private bool $hstsSubDomains = false;
+
+    private ResponseFactoryInterface $responseFactory;
+
+    public function __construct(ResponseFactoryInterface $responseFactory)
+    {
+        $this->responseFactory = $responseFactory;
+    }
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        if ($this->redirect && strcasecmp($request->getUri()->getScheme(), 'http') === 0) {
+            $url = (string)$request->getUri()->withScheme('https')->withPort($this->port);
```

Comment:
Would this overwrite port regardless if `$this->port` is set? Was it intended?

Reply:
If `$this->port` is `null` then URI will be without port like `https://my-site.org/index.php`. In this case Client will use default port for this protocol (443)

Otherwise the port will be part of the URI. And if we manually set the port value as 443 then URI will look like `https://my-site.org:443/index.php`